File: desafios/desafio001/BuscaOrdenacaoMVC_Java/src/view/MenuView.java
package view;

import controller.DadoController;
import java.util.List;
import java.util.Scanner;
import model.Dado;

/**
 * Classe responsável por exibir o menu e interagir com o usuário
 */
public class MenuView {
    private DadoController controller;

    Scanner scanner = new Scanner(System.in);

    /**
     * construtor da classe MenuView
     * @param controller instância do controlador de dados
     */
    public MenuView(DadoController controller) {
        this.controller = controller;
    }

    public void iniciar() {

        int opcao;

        do {
            exibirMenu();

            opcao = scanner.nextInt();
            scanner.nextLine();

            switch (opcao) {
                case 1:
                    carregarArquivo();
                    break;

                case 2:
                    mostrarDados();
                    break;

                case 3:
                    controller.bubbleSort();
                    mostrarDados("\nDados ordenados com Bubble Sort:");
                    break;

                case 4:
                    controller.quickSort();
                    mostrarDados("\nDados ordenados com Quick Sort:");
                    break;
                
                case 5:
                    buscarLinear();
                    break;
                
                case 6:
                    buscarBinaria();
                    break;

                case 7:
                    System.out.println("Encerrando...");
                    break;

                default:
                    System.out.println("Opção inválida!");

            }

        } while (opcao != 7);

    }

    
    private void exibirMenu() {
        System.out.println("--- MENU ---");
        System.out.println("1 - Carregar arquivo");
        System.out.println("2 - Mostrar dados");
        System.out.println("3 - Ordenar (Bubble Sort)");
        System.out.println("4 - Ordenar (Quick Sort)");
        System.out.println("5 - Buscar (Linear)");
        System.out.println("6 - Buscar (Binária)");
        System.out.println("7 - Sair");
        System.out.print("Escolha uma opção: ");

    }

    /**
     * carrega os dados de um arquivo de texto e armazena na lista de dados
     */
    private void carregarArquivo() {
        System.out.print("Informe o caminho do arquivo: ");
        String caminho = scanner.nextLine();

        if (controller.carregarArquivo(caminho)) {
            System.out.println("Arquivo carregado com sucesso!");
        } else {
            System.out.println("Erro ao ler o arquivo!");
        }

    }

    /**
     * mostra os dados carregados na tela
     */
    private void mostrarDados() {
        mostrarDados("\nDados carregados:");
    }

    private void mostrarDados(String titulo) {

        List<Dado> dados = controller.listarDados();

        if (dados.isEmpty()) {

            System.out.println("\nNenhum dado carregado");
            return;

        }

        System.out.println(titulo);

        for (Dado dado : dados) {
            System.out.println(dado);
        }

    }


    /**
     * método para buscar um valor na lista de dados utilizando busca linear
     */
    private void buscarLinear() {

    System.out.print("Informe o valor para busca linear: ");
    String valor = scanner.nextLine();

    Dado resultado = controller.buscaLinear(valor);

        if (resultado != null) {
            System.out.println("Valor encontrado: " + resultado.getValor());
        } else {
            System.out.println("Valor não encontrado");
        }

    }

    /**
     * método para buscar um valor na lista de dados utilizando busca binária
     */
    private void buscarBinaria() {

    System.out.print("Informe o valor para busca binária: ");
    String valor = scanner.nextLine();

    Dado resultado = controller.buscaBinaria(valor);

        if (resultado != null) {
            System.out.println("Valor encontrado: " + resultado.getValor());
        } else {
            System.out.println("Valor não encontrado");
        }

    }
}

File: desafios/desafio001/BuscaOrdenacaoMVC_CSharp/src/View/MenuView.cs
using BuscaOrdenacaoMVC_CSharp.Controller;
using BuscaOrdenacaoMVC_CSharp.Model;

namespace BuscaOrdenacaoMVC_CSharp.View;

public class MenuView
{
    private DadoController controller;

    public MenuView(DadoController controller)
    {
        this.controller = controller;
    }

    public void Iniciar()
    {
        int opcao;

        do
        {
            ExibirMenu();

            string? entrada = Console.ReadLine();

            if (!int.TryParse(entrada, out opcao))
            {
                Console.WriteLine("Entrada inválida! Digite um número");
                continue;
            }

            switch (opcao)
            {
                case 1:
                    CarregarArquivo();
                    break;

                case 2:
                    MostrarDados();
                    break;

                case 3:
                    controller.BubbleSort();
                    MostrarDados("\nDados ordenados com Bubble Sort:");
                    break;

                case 4:
                    controller.QuickSort();
                    MostrarDados("\nDados ordenados com Quick Sort:");
                    break;

                case 5:
                    BuscarLinear();
                    break;

                case 6:
                    BuscarBinaria();
                    break;

                case 7:
                    Console.WriteLine("Encerrando...");
                    break;

                default:
                    Console.WriteLine("Opção inválida!");
                    break;
            }

        } while (opcao != 7);
    }

    private void ExibirMenu()
    {
        Console.WriteLine("--- MENU ---");
        Console.WriteLine("1 - Carregar arquivo");
        Console.WriteLine("2 - Mostrar dados");
        Console.WriteLine("3 - Ordenar (Bubble Sort)");
        Console.WriteLine("4 - Ordenar (Quick Sort)");
        Console.WriteLine("5 - Buscar (Linear)");
        Console.WriteLine("6 - Buscar (Binária)");
        Console.WriteLine("7 - Sair");
        Console.Write("Escolha uma opção: ");
    }


    /// <summary>
    /// método que carrega os dados de um arquivo informado pelo usuário, 
    /// chamando o método CarregarArquivo do controller e exibindo mensagens de sucesso ou erro
    /// </summary>
    private void CarregarArquivo()
    {
        Console.Write("Informe o caminho do arquivo: ");
        string? caminho = Console.ReadLine();

        if (!string.IsNullOrWhiteSpace(caminho))
        {
            // chama o método CarregarArquivo do controller, passando o caminho informado pelo usuário
            if (controller.CarregarArquivo(caminho))
            {
                Console.WriteLine("Arquivo carregado com sucesso!");
            }
            else
            {
                Console.WriteLine("Erro ao ler o arquivo!");
            }
        }
        else
        {
            Console.WriteLine("Caminho inválido!");
        }
    }

    /// <summary>
    /// método que exibe os dados carregados
    /// </summary>
    /// <param name="titulo">título a ser exibido antes dos dados</param>
    private void MostrarDados(string titulo = "\nDados carregados:")
    {
        List<Dado> dados = controller.ListarDados();

        if (dados.Count == 0)
        {
            Console.WriteLine("\nNenhum dado carregado");
            return;
        }

        Console.WriteLine(titulo);

        foreach (Dado dado in dados)
        {
            Console.WriteLine(dado);
        }
    }

    /// <summary>
    /// método que realiza uma busca linear por um valor informado pelo usuário
    /// </summary>
    private void BuscarLinear()
    {
        Console.Write("Informe o valor para busca linear: ");
        string? valor = Console.ReadLine();

        // chama o método de busca linear do controller
        Dado? resultado = controller.BuscaLinear(valor ?? "");

        if (resultado != null)
        {
            Console.WriteLine("Valor encontrado: " + resultado.Valor);
        }
        else
        {
            Console.WriteLine("Valor não encontrado");
        }
    }

    /// <summary>
    /// método que realiza uma busca binária por um valor informado pelo usuário, 
    /// assumindo que os dados estão ordenados
    /// </summary>
    private void BuscarBinaria()
    {
        Console.Write("Informe o valor para busca binária: ");
        string? valor = Console.ReadLine();

        Dado? resultado = controller.BuscaBinaria(valor ?? "");

        if (resultado != null)
        {
            Console.WriteLine("Valor encontrado: " + resultado.Valor);
        }
        else
        {
            Console.WriteLine("Valor não encontrado");
        }
    }
}
